### PR TITLE
2FA change and tests

### PR DESCRIPTION
**Pull request by k-jell**
closes CRJI/EIC#605

### CODE REVIEW

File: testsuite/test_twofactor_change.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    assert response.url == '/accounts/totp/confirm/'
+    assert sum(1 for device in django_otp.devices_for_user(user['user'])) == 2
+    new_device = (TOTPDevice.objects
+                  .devices_for_user(user['user']).get(name='new_device_name'))
+    _reset_last_use(device)
+    confirm_response = client.post('/accounts/totp/confirm/', {
+        'new_token': _totp(new_device, now()),
+    })
+    assert confirm_response.context['success']
+    assert sum(1 for device in django_otp.devices_for_user(user['user'])) == 1
```
Comment (gabriel-v) [inline testsuite/test_twofactor_change.py]: > `     assert sum(1 for device in django_otp.devices_for_user(user['user'])) == 1`


`assert len(django_otp.devices_for_user(user['user'])) == 1`

or even better, move this to some `def user_device_count(user):` function up top in this module and use it inside your tests like this: `assert user_device_count(user) == 1`

File: testsuite/test_twofactor_change.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    client.login(username=user['username'], password=user['password'])
+    device = test_device(user=user['user'])
+    _reset_last_use(device)
+    response = client.post('/accounts/totp/change/', {
+        'password': user['password'],
+        'token': _totp(device, now()),
+        'new_name': 'new_device_name',
+    })
+    assert response.url == '/accounts/totp/confirm/'
+    assert sum(1 for device in django_otp.devices_for_user(user['user'])) == 2
```
Comment (gabriel-v) [inline testsuite/test_twofactor_change.py]: change to `assert user_device_count(user) == 2`, see below

File: testsuite/test_twofacter_remove.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+from django.utils.timezone import now
```
Comment (gabriel-v) [inline testsuite/test_twofacter_remove.py]: i would keep only one test module: `testsuite/test_twofactor.py` with the various tests (remove, add, etc) and the helper functions. We don't need this many files lying around, and the tests are related to one another (add, remove, edit same type of object) and should live next to each other. You will find that you've got some duplicated code that you can move into functions declared above the tests.

File: testsuite/test_twofactor_add.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    client.login(username=user['username'], password=user['password'])
+    device = test_device(user=user['user'])
+    _reset_last_use(device)
+    response = client.post('/accounts/totp/add/', {
+        'password': user['password'],
+        'token': _totp(device, now()),
+        'new_name': 'new_device_name',
+    })
+    assert response.url == '/accounts/totp/confirm/'
+    assert sum(1 for device in django_otp.devices_for_user(user['user'])) == 2
```
Comment (gabriel-v) [inline testsuite/test_twofactor_add.py]: same as below: `assert user_device_count(user) == 2`

File: testsuite/test_twofactor_change.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    new_device = (TOTPDevice.objects
+                  .devices_for_user(user['user']).get(name='new_device_name'))
+    _reset_last_use(device)
+    confirm_response = client.post('/accounts/totp/confirm/', {
```
Comment (gabriel-v) [inline testsuite/test_twofactor_change.py]: Use `reverse()` instead of hardcoded URL -- https://docs.djangoproject.com/en/3.2/ref/urlresolvers/#reverse

We don't want to change the tests every time we change the URLs.

File: testsuite/test_twofactor_change.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+from test_twofactor_invitation import _totp, _reset_last_use
+import django_otp
+
+
+def test_totp_change(client, test_user, test_device):
+    user = test_user()
+    client.login(username=user['username'], password=user['password'])
+    device = test_device(user=user['user'])
+    _reset_last_use(device)
+    response = client.post('/accounts/totp/change/', {
```
Comment (gabriel-v) [inline testsuite/test_twofactor_change.py]: Use `reverse()` instead of hardcoded URL -- https://docs.djangoproject.com/en/3.2/ref/urlresolvers/#reverse

We don't want to change the tests every time we change the URLs.

File: testsuite/test_twofactor_add.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+from django_otp.plugins.otp_totp.models import TOTPDevice
+import django_otp
+
+
+def test_totp_add(client, test_user, test_device):
+    user = test_user()
+    client.login(username=user['username'], password=user['password'])
+    device = test_device(user=user['user'])
+    _reset_last_use(device)
+    response = client.post('/accounts/totp/add/', {
```
Comment (gabriel-v) [inline testsuite/test_twofactor_add.py]: use reverse() here, see below

File: testsuite/test_twofactor_add.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        'password': user['password'],
+        'token': _totp(device, now()),
+        'new_name': 'new_device_name',
+    })
+    assert response.url == '/accounts/totp/confirm/'
```
Comment (gabriel-v) [inline testsuite/test_twofactor_add.py]: use reverse() here, see below

File: testsuite/test_twofacter_remove.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+import django_otp
+
+
+def test_totp_remove(client, test_user, test_device):
+    user = test_user()
+    client.login(username=user['username'], password=user['password'])
+    test_device(user=user['user'])
+    device2 = test_device(user=user['user'], name='device2')
+    _reset_last_use(device2)
+    response = client.post('/accounts/totp/remove/', {
```
Comment (gabriel-v) [inline testsuite/test_twofacter_remove.py]: use reverse(), see below

File: testsuite/test_profile.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
+def test_profile(client, test_user):
+    resp = client.get('/accounts/profile')
```
Comment (gabriel-v) [inline testsuite/test_profile.py]: use reverse(), see below

File: liquidcore/site/urls.py [hunk 11]
```diff
@@ -15,10 +25,25 @@
...
         path('invitation/<code>', twofactor_views.invitation),
+        path('accounts/totp/change/', twofactor_views.change_totp),
+        path('accounts/totp/confirm/',
+             twofactor_views.confirm_totp_change),
+        path('accounts/totp/settings/',
+             twofactor_views.totp_settings),
+        path('accounts/totp/add/',
+             twofactor_views.totp_add),
+        path('accounts/totp/remove/',
+             twofactor_views.totp_remove),
```
Comment (gabriel-v) [inline liquidcore/site/urls.py]: Add names to all of these, so you can use them with `reverse()` in the tests

File: testsuite/test_login.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+def test_login_page(client):
+    url = reverse('login')
+    resp = client.get(url)
+    assert resp.status_code == 200
+
+
+def test_login_no_totp(client, settings, test_user):
+    settings.LIQUID_2FA = False
+    user = test_user()
+    client.post('/accounts/login/', {
```
Comment (gabriel-v) [inline testsuite/test_login.py]: use reverse()

File: testsuite/test_login.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    assert is_logged_in(client)
+    client.logout()
+    assert not is_logged_in(client)
+
+
+def test_login_totp(client, test_user, test_device):
+    user = test_user()
+    device = test_device(user=user['user'])
+    _reset_last_use(device)
+    client.post('/accounts/login/', {
```
Comment (gabriel-v) [inline testsuite/test_login.py]: use reverse()

File: testsuite/test_login.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
+import pytest
+from django.urls import reverse
+from django.utils.timezone import now
+from test_twofactor_invitation import _totp, _reset_last_use, is_logged_in
```
Comment (gabriel-v) [inline testsuite/test_login.py]: Don't import functions from other tests. Test modules should be independent of one another, so you can skip and delete them with ease.

You should:
- put the helper functions like `totp` or `is logged in` in conftest without the `@pytest.fixture()` decorator
- collapse `_reset_last_use` into the `test_device` fixture, because you seem to always run this function as soon as the test starts and/or you need to do something with said device.
- `from .conftest import totp, is_logged_in` -- basically everything you want to use directly and is not a fixture

File: testsuite/test_admin.py [hunk 15]
```diff
@@ -0,0 +1,23 @@
+from django.utils.timezone import now
+from test_twofactor_invitation import _totp
```
Comment (gabriel-v) [inline testsuite/test_admin.py]: same as below: don't import from other tests, refactor

File: testsuite/conftest.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+                         .create_user(username=test_user_name,
+                                      password=test_user_pw)),
+                'password': test_user_pw,
+                'username': test_user_name,
+        }
+    return make_user
+
+
+@pytest.fixture
+def test_user_name():
```
Comment (gabriel-v) [inline testsuite/conftest.py]: Don't call any of your fixtures `test_*` because it creates confusion with the naming scheme of the actual tests. Not that it's a problem with the code, but it creates chaos inside our understanding of what's what.

You could call your fixtures `create_normal_user`, `create_admin_user` and `create_device`, since you wanted all of them to return functions that would let you create however many accounts you wanted. 

For the normal/admin users, you aren't using any of these multiple times, so you should simplify and return the User directly instead of returning a wrapping function. In those cases, you'd rename the fixtures to just `admin_user` and `normal_user` and use them directly as `normal_user['password']`

Also, don't use fixtures where you could settle for a simple constant 
```
NORMAL_USER_NAME = 'x'
NORMAL_USER_PASS = 'y'
ADMIN_USER_NAME = 'z'
ADMIN_USER_PASS = 't'
```

you don't have to prefix everything with `TEST_`, we know where we are

File: testsuite/test_admin.py [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+    response = client.post('/admin/', {
+        'username': 'testadmin',
+        'password': 'admin-pw',
+        'otp_token': _totp(device, now())
+    }, follow=True)
+    assert response.status_code == 200
+
+
+def test_admin_standard_user(client, django_user_model):
+    username = 'user1'
```
Comment (gabriel-v) [inline testsuite/test_admin.py]: Why not use the fixture to get a normal user here, and you instead duplicate that code?

File: testsuite/conftest.py [hunk 18]
```diff
@@ -0,0 +1,47 @@
+import pytest
+from liquidcore.twofactor import devices
+
+
+@pytest.fixture
+def test_user(django_user_model, test_user_name, test_user_pw):
+    def make_user():
+        return {
+                'user': (django_user_model.objects
```
Comment (gabriel-v) [inline testsuite/conftest.py]: Write comment here explaining why you can't just return the User instance (because there's no `User.get_password`) and you have to return a dict containing that User instance.

Also, you don't really need the `username` field here, right? You can just do `x['user'].get_username()`.

Since we're here: using dicts for fixed structures is ugly. You want to write `x.user.get_username()` and `x.password`, not the ugly `[""]` -- so you can do:
- simple `class`: need to write `__init__`, too long, too lazy
- https://docs.python.org/3/library/collections.html#collections.namedtuple
- the new data classes: https://docs.python.org/3/library/dataclasses.html -- but are we running python 3.7 here, or 3.6?

I'd go with the named tuple

File: liquidcore/twofactor/views.py [hunk 19]
```diff
@@ -40,3 +49,109 @@ def invitation(request, code):
...
+                            password=request.POST['password']):
+            bad_password = True
+
+        if not (bad_password or bad_token):
+            new_device = add_device(request.user, request.POST['new_name'])
+            request.session['new_device'] = new_device.id
+
+            request.session['otp_png'] = get_png(request.user, new_device)
+
+            return redirect(confirm_totp_change)
```
Comment (gabriel-v) [inline liquidcore/twofactor/views.py]: I don't like this redirect here -- why do we need a separate page for confirmation? The whole thing is getting too complex, with users being able to finalize the first page and create some device, but then forget to confirm the next form and end up with devices that aren't activated / verified / saved onto their phone.

Please collapse both of these into a single form that does new device creation and confirmation too, and remove the `confirm_totp_change` page and template. You visit the "change TOTP" page, you put in the password, the old code, the old token, scan the QR, put in the new token, select checkbox if you want to remove everything else or not, hit OK, done.

File: testsuite/test_admin.py [hunk 20]
```diff
@@ -0,0 +1,23 @@
+from django.utils.timezone import now
+from test_twofactor_invitation import _totp
+
+
+def test_admin_view(client, test_admin, test_device):
+    admin = test_admin()
+    device = test_device(user=admin)
+    response = client.post('/admin/', {
```
Comment (gabriel-v) [inline testsuite/test_admin.py]: Rename this into `test_admin_page_login_with_totp` to be more descriptive.

It's kinda pointless testing this without testing any failure cases. Before doing the correct login, try:
- bad password = no login
- bad TOTP = no login
- bad username = no login

here, look at these guys running a bank: https://twitter.com/esraefe/status/1397842160607711232

File: testsuite/test_login.py (hunk 12 above)
Comment (gabriel-v) [reply testsuite/test_login.py]: Also, see above: test bad username, test bad password.

File: testsuite/test_login.py [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def test_login_totp(client, test_user, test_device):
+    user = test_user()
+    device = test_device(user=user['user'])
+    _reset_last_use(device)
+    client.post('/accounts/login/', {
+        'username': user['username'],
+        'password': user['password'],
+        'otp_token': _totp(device, now())
```
Comment (gabriel-v) [inline testsuite/test_login.py]: Also, test that using bad username, password and OTP token means you don't get the login.

Test that TOTP with wrong date `(now() + timedelta(minutes=5))` also fails. You never know when the TOTP library gets social engineered, abandons normal functioning and just returns the QR code of `password` every time

File: testsuite/test_password_change.py [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+from django.utils.timezone import now
+from test_twofactor_invitation import _totp, _reset_last_use, is_logged_in
+
+
+def test_password_change_no_totp(settings, client, test_user):
+    settings.LIQUID_2FA = False
+    url = reverse('password_change')
+    user = test_user()
+    client.login(username=user['username'], password=user['password'])
+    client.post(url, {
```
Comment (gabriel-v) [inline testsuite/test_password_change.py]: Also check in this test that sending wrong old password and/or failing to duplicate the new password ends up with an error, before you do the actual password change.

File: testsuite/test_password_change.py [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+    user = test_user()
+    device = test_device(user=user['user'])
+    _reset_last_use(device)
+    client.login(username=user['username'], password=user['password'])
+    assert is_logged_in(client)
+    client.post(url, {
+        'old_password': user['password'],
+        'new_password1': 'new_password',
+        'new_password2': 'new_password',
+        'otp_token': _totp(device, now())
```
Comment (gabriel-v) [inline testsuite/test_password_change.py]: Just try to change password with a bad TOTP token for this one. Checking old password and new passwords goes throu same code path as the test above, right?

File: liquidcore/twofactor/templates/totp-remove.html [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+  </fieldset>
+
+  <button type="submit" class="btn btn-primary"> 
+    Remove
+  </button>
+  <a href="/" class="btn btn-secondary">Cancel</a>
+</form>
+
+  {% if bad_token %}
+    <p>Invalid token, please try again.</p>
```
Comment (gabriel-v) [inline liquidcore/twofactor/templates/totp-remove.html]: put some "error" class coloring on these `<p>`s:
https://getbootstrap.com/docs/3.3/css/#helper-classes-backgrounds

you can do `<p class="bg-warning">` and `bg-success` respectively

File: liquidcore/twofactor/templates/totp-settings.html [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+
+{% block content %}
+
+<h1>2FA device settings</h1>
+
+  <h2>Currently registered devices:</h2>
+  <p><b>Note:</b> The device named like your username is the device you registered when setting up the two factor authentication</p>
+  <ul>
+  {% for device in devices %}
+  <li>{{ device.name }}</li>
```
Comment (gabriel-v) [inline liquidcore/twofactor/templates/totp-settings.html]: What fields do we have here? Please explore.

We'd want at least the date when it was added, and also the date for the last login.

Now that I look at the code: https://django-otp-official.readthedocs.io/en/stable/_modules/django_otp/models.html I see it doesn't have any of the timestamps we wanted -- I'll open a separate issue about tracking when / how TOTP devices were named.

File: liquidcore/twofactor/templates/totp-remove.html [hunk 26]
```diff
@@ -0,0 +1,57 @@
+{% extends "auth_base.html" %}
+
+{% block content %}
+
+<h1>Remove 2FA devices</h1>
+
+<form method="post">
+  {% csrf_token %}
+  <p>Enter your password and a valid OTP token to remove all devices but one. Choose the device you want to keep from the dropdown menu.</p>
```
Comment (gabriel-v) [inline liquidcore/twofactor/templates/totp-remove.html]: With this logic, it's easy to lock yourself out by picking the wrong key to remove from the drop-down.

Do we make sure the TOTP token entered is valid for the remaining device selected from the drop-down?

If we do it that way, then reword this message to sound less threatening. If we don't do it that way, then please change the TOTP removal function so it's impossible to lock yourself out.

Please expand the TOTP removal test with this case where you set the drop-down to some other device than the one that generated the token.

We could even do it simpler: we'd remove all the devices except for the one that had a correct code, assuming they're all going to have different secrets. That way, they have no drop down to pick from.

File: liquidcore/twofactor/templates/totp-settings.html [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+  {% endfor %}
+  </ul>
+
+  <a href="/accounts/totp/change/" class="btn btn-secondary">Change device</a>
```
Comment (gabriel-v) [inline liquidcore/twofactor/templates/totp-settings.html]: make the "create" button `btn-primary` and the "remove" button `btn-danger` instead of having them all the same shade of blue

File: testsuite/test_password_change.py [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+    url = reverse('password_change')
+    user = test_user()
+    client.login(username=user['username'], password=user['password'])
+    client.post(url, {
+        'old_password': user['password'],
+        'new_password1': 'new_password',
+        'new_password2': 'new_password',
+    })
+    client.logout()
+    client.login(username=user['username'], password='new_password')
```
Comment (gabriel-v) [inline testsuite/test_password_change.py]: After logging out, attempt to login with the old password and verify it fails.

File: testsuite/test_password_change.py [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+    client.login(username=user['username'], password=user['password'])
+    assert is_logged_in(client)
+    client.post(url, {
+        'old_password': user['password'],
+        'new_password1': 'new_password',
+        'new_password2': 'new_password',
+        'otp_token': _totp(device, now())
+    })
+    client.logout()
+    client.login(username=user['username'], password='new_password')
```
Comment (gabriel-v) [inline testsuite/test_password_change.py]: So how exactly do you login here without the TOTP?

File: testsuite/test_login.py [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+    assert is_logged_in(client)
+    client.logout()
+    assert not is_logged_in(client)
+
+
+def test_homepage(client, test_user):
+    resp = client.get('')
+    assert resp.status_code == 302
+    user = test_user()
+    client.login(username=user['username'], password=user['password'])
```
Comment (gabriel-v) [inline testsuite/test_login.py]: The docs for the django client login function are here: https://docs.djangoproject.com/en/3.2/topics/testing/tools/#django.test.Client.login

Question: Why does this not fail, since we have TOTP = on in this test? I think we're missing some configuration to make it mandatory to have `otp_token` set here too, on `authenticate()` and `client.login()`. If that's the case, please expand this test to check that missing the OTP token from the auth request fails the login.

**Review by gabriel-v — CHANGES_REQUESTED** [review]
A great start! Lots of small changes to make around the templates and the tests.

I want to have a single form for any step, so please remove the "TOTP change confirmation" page as mentioned in the comments below. Any operation should start and end on a single page; we can't assume our users will successfully visit all the pages you want them to, every single time.

**Review by gabriel-v — APPROVED** [review]
LGTM, will need some extensive testing for this, but I don't see any major issues with the code.
